src/protocols/bdx/tests/TestBdxTransferSession.cpp

```diff
@@ -0,0 +1,335 @@
+#include <protocols/bdx/BdxMessages.h>
+#include <protocols/bdx/BdxTransferSession.h>
+
+#include <nlunit-test.h>
+
+#include <support/CodeUtils.h>
+#include <support/UnitTestRegistration.h>
+
+using namespace ::chip;
+using namespace ::chip::BDX;
+
+void SendAndVerifyQuery(nlTestSuite * inSuite, void * inContext, TransferSession & sender, TransferSession & receiver,
+                        TransferSession::OutputEvent & outEvent)
+{
+    // Verify that receiver emits BlockQuery message
+    CHIP_ERROR err = receiver.PrepareBlockQuery();
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    receiver.PollOutput(outEvent);
+    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::kOutput_MsgToSend);
+    System::PacketBufferHandle queryMsgToSend = outEvent.MsgData.Retain();
+    NL_TEST_ASSERT(inSuite, !queryMsgToSend.IsNull());
+
+    // Pass BlockQuery to sender
+    err = sender.HandleMessageReceived(queryMsgToSend.Retain());
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+
+    // Verify sender emits QueryReceived event
+    sender.PollOutput(outEvent);
+    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::kOutput_QueryReceived);
+}
+
+void SendAndVerifyArbitraryBlock(nlTestSuite * inSuite, void * inContext, TransferSession & sender, TransferSession & receiver,
+                                 TransferSession::OutputEvent & outEvent, bool isEof)
+{
+    CHIP_ERROR err                         = CHIP_NO_ERROR;
+    uint16_t maxBlockSize                  = sender.GetTransferBlockSize();
+    System::PacketBufferHandle fakeDataBuf = System::PacketBuffer::NewWithAvailableSize(maxBlockSize);
+    if (fakeDataBuf.IsNull())
+    {
+        NL_TEST_ASSERT(inSuite, false);
+        return;
+    }
+
+    uint8_t * fakeBlockData = fakeDataBuf->Start();
+
+    // Provide Block data and verify sender emits Block message
+    TransferSession::BlockData blockData;
+    blockData.Data   = fakeBlockData;
+    blockData.Length = maxBlockSize;
+    blockData.IsEof  = isEof;
+
+    err = sender.PrepareBlock(blockData);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    sender.PollOutput(outEvent);
+    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::kOutput_MsgToSend);
+    System::PacketBufferHandle blockMsgToSend = outEvent.MsgData.Retain();
+    NL_TEST_ASSERT(inSuite, !blockMsgToSend.IsNull());
+
+    // Pass Block message to receiver and verify Block is received
+    err = receiver.HandleMessageReceived(blockMsgToSend.Retain());
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    receiver.PollOutput(outEvent);
+    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::kOutput_BlockReceived);
+    System::PacketBufferHandle receivedBlock = outEvent.MsgData.Retain();
+    NL_TEST_ASSERT(inSuite, !receivedBlock.IsNull());
+    NL_TEST_ASSERT(inSuite, outEvent.blockdata.Data != nullptr);
```

Comment:
Could you verify that the size and data match by generating some usable fake data, such as incrementing counter bytes in the payload? That would ensure that data extraction is satisfactory

Reply:
Sure, like having a static variable that's incremented each time and writing that to the first byte in the payload?